File: rust/cyclotron-node/examples/dequeue_ordering.js
const assert = require('assert');
const cyclotron = require('../.');
const crypto = require('crypto');


// This file (and the rust equivalent) are me trying to figure out why, sometimes, the order of jobs returned
// from the queue does not match the priority order set previous (BUT ONLY IN NODE). I've been unable to reproduce
// this in rust, so the error probably lies in some node binding gremlins, but I have /not even close/ to enough
// knowledge of the node runtime to figure out why that might be the case. I've checked, and the priority is being
// set correctly in the DB, it's just ??? not being used ??? for the subsequent dequeue call ??? sometimes ???.

// UPDATE - this file now also runs without issues... I don't know what's going on, but putting it down to
// "oliver needs to stop writing code at 1am"... Gonna go sketch out some management command stuff now.

const JOB_STATES = Object.freeze({
    AVAILABLE: "available",
    RUNNING: "running",
    FAILED: "failed",
    COMPLETED: "completed",
});

const AVAILABLE_WORKERS = Object.freeze({
    FETCH: "fetch",
    HOG: "hog",
});

async function main() {
    let poolConfig = {
        host: 'localhost',
        port: 5432,
        user: 'posthog',
        password: 'posthog',
        db: 'posthog'
    }

    let managerConfig = {
        shards: [poolConfig]
    }

    // Most processes will only need to do one of these, but we can do both here for demonstration purposes
    await cyclotron.init_worker(JSON.stringify(poolConfig))
    await cyclotron.init_manager(JSON.stringify(managerConfig))

    // Maybe inits won't throw on re-calling, and are also short-circuiting to be almost free, so safe to call frequently
    // (although I still wouldn't call them in a loop)
    await cyclotron.maybe_init_worker(JSON.stringify(poolConfig))
    await cyclotron.maybe_init_manager(JSON.stringify(managerConfig))

    let five_mintes_ago = new Date(new Date().getTime() - 5 * 60000).toISOString()
    let queue_name = "default"

    let job_1 = {
        team_id: 1,
        waiting_on: AVAILABLE_WORKERS.FETCH,
        queue_name: queue_name,
        priority: 0,
        scheduled: five_mintes_ago,
        function_id: crypto.randomUUID(), // Is nullable
        vm_state: null,
        parameters: null,
        metadata: null
    }

    let job_2 = {
        team_id: 1,
        waiting_on: AVAILABLE_WORKERS.FETCH,
        queue_name: queue_name,
        priority: 1,
        scheduled: five_mintes_ago,
        function_id: crypto.randomUUID(), // Is nullable
        vm_state: null,
        parameters: null,
        metadata: null
    }

    let start = new Date().getTime()
    let count = 0;
    while(true) {

        await cyclotron.create_job(JSON.stringify(job_1))
        await cyclotron.create_job(JSON.stringify(job_2))

        // Jobs (as well as any other 'complex' data shape) are serialized across the API boundary,
        // because that's (according to the neon maintainers) /actually faster/ than doing a bunch
        // of cross-runtime pointer chasing.
        let jobs = JSON.parse(await cyclotron.dequeue_jobs(queue_name, AVAILABLE_WORKERS.FETCH, 2))
        assert(jobs.length === 2)
        assert(jobs[0].function_id === job_1.function_id)
        assert(jobs[1].function_id === job_2.function_id)

        job_1 = jobs[0]
        job_2 = jobs[1]

        cyclotron.set_state(job_1.id, JOB_STATES.AVAILABLE)
        cyclotron.set_state(job_2.id, JOB_STATES.AVAILABLE)

        cyclotron.set_priority(job_1.id, 2)
        cyclotron.set_priority(job_2.id, 1)

        await cyclotron.flush_job(job_1.id)
        await cyclotron.flush_job(job_2.id)

        jobs = JSON.parse(await cyclotron.dequeue_with_vm_state("default", AVAILABLE_WORKERS.FETCH, 2))

        // Now, since we've re-ordered the jobs, we should expect job_2 to be returned first
        assert(jobs.length === 2)
        assert(jobs[0].id == job_2.id)
        assert(jobs[1].id == job_1.id)

        // Now we'll mark these jobs as completed
        cyclotron.set_state(job_1.id, JOB_STATES.COMPLETED)
        cyclotron.set_state(job_2.id, JOB_STATES.COMPLETED)

        // And flush them back to the queue
        await cyclotron.flush_job(job_1.id)
        await cyclotron.flush_job(job_2.id)

        count++;
        if (count % 10 == 0) {
            let elapsed = new Date().getTime() - start
            console.log(`Looped ${count} jobs in ${elapsed}ms)`)
        }
        if(count > 2000) {
            break;
        }
    }

}

main()

File: rust/cyclotron-core/src/bin/dequeue_ordering.rs
use chrono::{Duration, Utc};
use cyclotron_core::{
    base_ops::{JobInit, JobState, WaitingOn},
    manager::{ManagerConfig, QueueManager},
    worker::Worker,
    PoolConfig,
};
use uuid::Uuid;

// See the comment at the top of deqeueu_ordering.js in the node libraries examples folder for
// a description of what this is for. This function happily runs forever, but the exact equivalent
// in javascript fails almost immediately, and I do not know why.
#[tokio::main]
async fn main() {
    let pool_config = PoolConfig {
        host: "localhost".to_string(),
        port: 5432,
        user: "posthog".to_string(),
        password: "posthog".to_string(),
        db: "posthog".to_string(),
        max_connections: None,
        min_connections: None,
        acquire_timeout_seconds: None,
        max_lifetime_seconds: None,
        idle_timeout_seconds: None,
    };

    let manager_config = ManagerConfig {
        shards: vec![pool_config.clone()],
    };

    let manager = QueueManager::new(manager_config).await.unwrap();
    let worker = Worker::new(pool_config).await.unwrap();

    let now = Utc::now() - Duration::minutes(1);
    let start = Utc::now();
    let mut count = 0;
    loop {
        let test_job = JobInit {
            team_id: 1,
            waiting_on: WaitingOn::Fetch,
            queue_name: "default".to_string(),
            priority: 0,
            scheduled: now,
            function_id: Some(Uuid::now_v7()),
            vm_state: None,
            parameters: None,
            metadata: None,
        };

        let mut test_job_2 = test_job.clone();
        test_job_2.priority = 1;
        let test_job_2 = test_job_2;

        manager.create_job(test_job).await.unwrap();
        manager.create_job(test_job_2).await.unwrap();

        let jobs = worker
            .dequeue_jobs("default", WaitingOn::Fetch, 2)
            .await
            .unwrap();

        assert!(jobs.len() == 2);
        assert!(jobs[0].priority == 0);
        assert!(jobs[1].priority == 1);

        let job_1_id = jobs[0].id;
        let job_2_id = jobs[1].id;

        worker.set_state(jobs[0].id, JobState::Available).unwrap();
        worker.set_state(jobs[1].id, JobState::Available).unwrap();

        // Set job 1 to have the same priority as job 2 originally had
        worker.set_priority(jobs[0].id, 2).unwrap();
        // And set job 2 to have a priority higher than job 1
        worker.set_priority(jobs[1].id, 1).unwrap();

        worker.flush_job(jobs[0].id).await.unwrap();
        worker.flush_job(jobs[1].id).await.unwrap();

        let jobs = worker
            .dequeue_with_vm_state("default", WaitingOn::Fetch, 2)
            .await
            .unwrap();

        assert!(jobs.len() == 2);
        // Assert our priority re-ordering was respected
        assert!(jobs[0].id == job_2_id);
        assert!(jobs[1].id == job_1_id);

        // mark these as completed
        worker.set_state(jobs[0].id, JobState::Completed).unwrap();
        worker.set_state(jobs[1].id, JobState::Completed).unwrap();

        worker.flush_job(jobs[0].id).await.unwrap();
        worker.flush_job(jobs[1].id).await.unwrap();

        count += 1;
        println!("Elapsed: {:?}, count: {}", Utc::now() - start, count);
    }
}
